### ISSUE TITLE
Expression evaluation with null

### ISSUE THREAD

**Issue by ciolben** [pr83/ylc#6]
Hello!

I have encoutered some difficulties to make the evaluation of this expression work correctly when I test for null:
```html
<span data-ylcLoop="port: ..." data-ylcEvents="click: ..." data-ylcBind="css.background: content.b==null ? 'antiquewhite' : 'white'">
```

This:
```javascript
content.b==null
```
is always evaluated to `false`, even if `content.b` is `null`.

Any ideas?
Thank you!

**Comment by telesma** [comment]
have you tried `content.b === null`

**Comment by ciolben** [comment]
Yes I did.

**Comment by pr83** [comment]
Hello,
It's a genuine bug and you are the first one who noticed that. Thanks.

I know what's causing it and it should be easy to fix, but before I cut a new release, I need to do some regression testing. I plan to cut a new release by tomorrow.

Until then, would it be an option for you to use this test instead:

`content.b == undefined`

**Comment by ciolben** [comment]
It is okay, I use another test as a workaround, but it is not as clean as doing the equality with null.
Thank you very much for taking care of that.

Have a nice day, I'm waiting for the update!

**Comment by pr83** [comment]
OK, so the release is cut and should be fixing this issue. Please let me know if it works for you.

**Comment by ciolben** [comment]
Thank you for the release!
However, after upgrading to 99.19, the issue is still there somehow. Let me explain.

It works now for tests with properties on the main object only (example `model.a`), but not if I have at least 2 levels of hierarchy like `model.a.b`.
Expression tests on `a.b==null` still always evaluates to `false`.

Test case (updated case from issue #5 , which has now both issues (useful for testing)):
View:
```html
<div id="hook">
	  <button data-ylcEvents="click: clickGetResult">GetResult</button>
	  <button data-ylcEvents="click: clickTestNull">TestNull</button>
		<table border="1">
		  <thead>
			<tr>
			  <th>Name</th>
			  <th>Content</th>
			</tr>
		  </thead>
		  <tbody>
			<tr data-ylcLoop="device: result">
			  <td data-ylcBind="text <- device.name "></td>
			  <td>
				<span data-ylcIf="!device.expanded" data-ylcEvents="click: clickRow(device)">+</span>
				<table data-ylcIf="device.expanded">
				  <thead>
					<tr>
					  <th><span data-ylcEvents="click: clickRow(device)">-</span>P1</th>
					  <th>P2</th>
					</tr>
				  </thead>
				  <tbody>
					<tr data-ylcLoop="content: device.contents">
					  <td data-ylcBind="text <- content.p1; css.background-color: testnull==null?'green':'red'"></td>
					  <td data-ylcBind="text <- content.p2; css.background-color: subobj.testnull==null?'green':'red'"></td>
					</tr>
				  </tbody>
				</table>
			  </td>
			</tr>
		  </tbody>
		</table>
	</div>
```
Controller:
```javascript
var controller = {
        init: function(model) {
          model.testnull = null;
          model.subobj = {};
          model.subobj.testnull = null;
          
          model.result = [
					{
						name: "Test 1",
						contents: [
							{
								p1: "test 1 a",
								p2: "test 1 b"
							}
						]
					},
					{
						name: "Test 2",
						contents: [
							{
								p1: "test 2 a",
								p2: "test 2 b"
							}
						]
					}
				  ];
        },
        clickRow: function(model, context, device) {
  				if (typeof device.expanded === 'undefined') {
  					device.expanded = false;
  				}
  				device.expanded = !device.expanded;
			  },
			  clickTestNull: function(model, context) {
			    if (model.subobj.testnull == null) {
			      model.subobj.testnull = 1;
			    } else {
			      model.subobj.testnull = null;
			    }
			    if (model.testnull == null) {
			      model.testnull = 1;
			    } else {
			      model.testnull = null;
			    }
			    
			  },
			  clickGetResult: function(model, context) {
			    context.updateModel(function(model){
  			    model.result = [
  					{
  						name: "Test 3",
  						contents: [
  							{
  								p1: "test 1 a",
  								p2: "test 1 b"
  							}
  						]
  					},
  					{
  						name: "Test 4",
  						contents: [
  							{
  								p1: "test 2 a",
  								p2: "test 2 b"
  							}
  						]
  					}
  				  ];
			    });
			  }
    }
```

**Comment by pr83** [comment]
I have realized there's still something I'm not getting. Please clarify.

Just to be sure we are on the same page, this is the HTML to use:

<!DOCTYPE html>
<html>

	<head>
		<script src="http://code.jquery.com/jquery-1.11.1.min.js"></script>
		<script src="yellowCode.js"></script>
	</head>

	<body>
		<div id="hook">
		  <button data-ylcEvents="click: clickGetResult">GetResult</button>
			<table border="1">
			  <thead>
				<tr>
				  <th>Name</th>
				  <th>Content</th>
				</tr>
			  </thead>
			  <tbody>
				<tr data-ylcLoop="device: result">
				  <td data-ylcBind="text <- device.name "></td>
				  <td>
					<span data-ylcIf="!device.expanded" data-ylcEvents="click: clickRow(device)">+</span>
					<table data-ylcIf="device.expanded">
					  <thead>
						<tr>
						  <th><span data-ylcEvents="click: clickRow(device)">-</span>P1</th>
						  <th>P2</th>
						</tr>
					  </thead>
					  <tbody>
						<tr data-ylcLoop="content: device.contents">
						  <td data-ylcBind="text <- content.p1; css.background-color: testnull==null?'green':'red'"></td>
						  <td data-ylcBind="text <- content.p2; css.background-color: subobj.testnull==null?'green':'red'"></td>
						</tr>
					  </tbody>
					</table>
				  </td>
				</tr>
			  </tbody>
			</table>
		</div>

		<script>

			var controller = {
				init: function(model) {
				  model.testnull = null;
				  model.subobj = {};
				  model.subobj.testnull = null;
				  
				  model.result = [
							{
								name: "Test 1",
								contents: [
									{
										p1: "test 1 a",
										p2: "test 1 b"
									}
								]
							},
							{
								name: "Test 2",
								contents: [
									{
										p1: "test 2 a",
										p2: "test 2 b"
									}
								]
							}
						  ];
				},
				
				clickRow: function(model, context, device) {
					if (typeof device.expanded === 'undefined') {
						device.expanded = false;
					}
					device.expanded = !device.expanded;
				},
				clickGetResult: function(model, context) {
					context.updateModel(function(model){
					model.result = [
						{
							name: "Test 3",
							contents: [
								{
									p1: "test 3 a",
									p2: "test 3 b"
								}
							]
						},
						{
							name: "Test 4",
							contents: [
								{
									p1: "test 4 a",
									p2: "test 4 b"
								}
							]
						}
					  ];
					});
				}
			}

			$("#hook").yellowCode(controller);

		</script>

	</body>

</html>

What I did:

- load the page
  - "Test 1" and "Test 2" appear in the left table column, the right column cells are collapsed
- manually expand the right columns
  - right column cells contain "test 1 a test 1 b" and "test 2 a test 2 b"
- click "GetResult"
  - "Test 3" and "Test 4" appear in the left table column, the right column cells are collapsed
- manually expand the right columns
  - cells contain "test 3 a test 3 b" and "test 4 a test 4 b"

Is this the correct sequence of steps? Are the results expected?

Thanks very much.

**Comment by ciolben** [comment]
I have the controller in a separated file that's why I show you the view and controller code in two different blocks.
Here is the complete code:
```html
<head>
	<script src="http://code.jquery.com/jquery-1.11.1.min.js"></script>
	<script src="yellowCode.js"></script>
</head>

<body>
	<div id="hook">
	  <button data-ylcEvents="click: clickGetResult">GetResult</button>
      <button data-ylcEvents="click: clickTestNull">TestNull</button>
		<table border="1">
		  <thead>
			<tr>
			  <th>Name</th>
			  <th>Content</th>
			</tr>
		  </thead>
		  <tbody>
			<tr data-ylcLoop="device: result">
			  <td data-ylcBind="text <- device.name "></td>
			  <td>
				<span data-ylcIf="!device.expanded" data-ylcEvents="click: clickRow(device)">+</span>
				<table data-ylcIf="device.expanded">
				  <thead>
					<tr>
					  <th><span data-ylcEvents="click: clickRow(device)">-</span>P1</th>
					  <th>P2</th>
					</tr>
				  </thead>
				  <tbody>
					<tr data-ylcLoop="content: device.contents">
					  <td data-ylcBind="text <- content.p1; css.background-color: testnull==null?'green':'red'"></td>
					  <td data-ylcBind="text <- content.p2; css.background-color: subobj.testnull==null?'green':'red'"></td>
					</tr>
				  </tbody>
				</table>
			  </td>
			</tr>
		  </tbody>
		</table>
	</div>

	<script>

		var controller = {
			init: function(model) {
			  model.testnull = null;
			  model.subobj = {};
			  model.subobj.testnull = null;
			  
			  model.result = [
						{
							name: "Test 1",
							contents: [
								{
									p1: "test 1 a",
									p2: "test 1 b"
								}
							]
						},
						{
							name: "Test 2",
							contents: [
								{
									p1: "test 2 a",
									p2: "test 2 b"
								}
							]
						}
					  ];
			},
			
			clickRow: function(model, context, device) {
				if (typeof device.expanded === 'undefined') {
					device.expanded = false;
				}
				device.expanded = !device.expanded;
			},
            clickTestNull: function(model, context) {
			    if (model.subobj.testnull == null) {
			      model.subobj.testnull = 1;
			    } else {
			      model.subobj.testnull = null;
			    }
			    if (model.testnull == null) {
			      model.testnull = 1;
			    } else {
			      model.testnull = null;
			    }
			    
			},
			clickGetResult: function(model, context) {
				context.updateModel(function(model){
				model.result = [
					{
						name: "Test 3",
						contents: [
							{
								p1: "test 3 a",
								p2: "test 3 b"
							}
						]
					},
					{
						name: "Test 4",
						contents: [
							{
								p1: "test 4 a",
								p2: "test 4 b"
							}
						]
					}
				  ];
				});
			}
		}

		$("#hook").yellowCode(controller);

	</script>

</body>
```
You're almost right. We can get the issues in different ways, the **quickest** is (moslty like you did):

The issue #5 :
  - load the page
    - "Test 1" and "Test 2" appear in the left table column, the right column cells are collapsed
  - click "GetResult"
    - "Test 3" and "Test 4" appear in the left table column, the right column cells are collapsed
  - manually expand the right columns
    - _doesn't expand_!

This issue:
  - load the page
    - "Test 1" and "Test 2" appear in the left table column, the right column cells are collapsed
  - manually expand the right columns
    - right column cells contain "test 1 a" in green, "test 1 b" in red and "test 2 a" in green, "test 2 b" in red
           _Both "test x b" should be green_!
  - click "TestNull" any number of time
    - _"test 1 b" and "test 2 b" stay red when they should be green_!

Is it clearer? (sorry if I was a bit confusing)

**Comment by pr83** [comment]
Hello,
Thanks very much for your comments.

Regarding this issue (null in model), I have made additional fixes and cut a new release 0.99.20. Please check.

Regarding the ylcIf, I will add a note to the original ticket shortly.

**Comment by ciolben** [comment]
Thank you for the fix, this time the tests on `null` are correct!
